File: GEULGWI-BACKEND/src/main/java/com/lovedbug/geulgwi/core/domain/curation/mapper/CurationMapper.java
package com.lovedbug.geulgwi.core.domain.curation.mapper;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import com.lovedbug.geulgwi.core.domain.book.dto.PopularBookResponse;
import com.lovedbug.geulgwi.core.domain.book.entity.Book;
import com.lovedbug.geulgwi.core.domain.curation.constant.CurationType;
import com.lovedbug.geulgwi.core.domain.curation.projection.CurationBook;
import com.lovedbug.geulgwi.core.domain.curation.dto.response.CurationContentResponse;
import com.lovedbug.geulgwi.core.domain.curation.dto.response.CurationItemResponse;
import com.lovedbug.geulgwi.core.domain.curation.projection.CurationQuote;
import com.lovedbug.geulgwi.core.domain.quote.dto.response.QuoteWithBookResponse;
import com.lovedbug.geulgwi.external.book_provider.aladdin.dto.AladdinBookResponse;

public class CurationMapper {

    public static CurationContentResponse toCurationContent(QuoteWithBookResponse quoteWithBook) {
        return CurationContentResponse.builder()
            .bookId(quoteWithBook.getBookId())
            .bookTitle(quoteWithBook.getBookTitle())
            .quoteId(quoteWithBook.getQuoteId())
            .author(quoteWithBook.getAuthor())
            .publisher(quoteWithBook.getPublisher())
            .imageName(quoteWithBook.getQuoteImageName())
            .build();
    }

    public static List<CurationContentResponse> toCurationContentListFromQuotes(List<QuoteWithBookResponse> quoteWithBooks) {
        return quoteWithBooks.stream()
            .map(CurationMapper::toCurationContent)
            .toList();
    }

    public static CurationContentResponse toCurationContent(AladdinBookResponse aladdinBook) {
        return CurationContentResponse.builder()
            .bookTitle(aladdinBook.getTitle())
            .author(aladdinBook.getAuthor())
            .publisher(aladdinBook.getPublisher())
            .bookCoverUrl(aladdinBook.getCoverUrl())
            .build();
    }

    public static CurationContentResponse toCurationContent(PopularBookResponse popularBook) {
        return CurationContentResponse.builder()
            .bookId(popularBook.bookId())
            .bookTitle(popularBook.bookTitle())
            .author(popularBook.author())
            .publisher(popularBook.publisher())
            .bookCoverUrl(popularBook.bookCoverUrl())
            .build();
    }

    public static List<CurationContentResponse> toCurationContentListFromBooks(List<PopularBookResponse> popularBooks) {
        return popularBooks.stream()
            .map(CurationMapper::toCurationContent)
            .toList();
    }

    public static CurationContentResponse toCurationContent(CurationBook curationBook) {
        return CurationContentResponse.builder()
            .bookId(curationBook.getBookId())
            .bookTitle(curationBook.getBookTitle())
            .author(curationBook.getAuthor())
            .publisher(curationBook.getPublisher())
            .bookCoverUrl(curationBook.getBookCoverUrl())
            .build();
    }

    public static List<CurationItemResponse> toCurationItemDtoListFromBooks(List<CurationBook> curationBooks) {
        Map<Long, List<CurationBook>> curationBooksByItemId = curationBooks.stream()
            .collect(Collectors.groupingBy(CurationBook::getCurationItemId));

        return curationBooksByItemId.entrySet().stream()
            .map(curationItemGroup -> {
                Long curationItemId = curationItemGroup.getKey();
                List<CurationBook> curationItems = curationItemGroup.getValue();

                CurationBook representativeBook = curationItems.get(0);

                List<CurationContentResponse> contents = curationItems.stream()
                    .map(CurationMapper::toCurationContent)
                    .toList();

                return new CurationItemResponse(
                    curationItemId,
                    representativeBook.getTitle(),
                    representativeBook.getDescription(),
                    CurationType.BOOK,
                    contents
                );
            })
            .toList();
    }

    public static CurationContentResponse toCurationContent(CurationQuote curationQuote) {
        return CurationContentResponse.builder()
            .bookId(curationQuote.getBookId())
            .bookTitle(curationQuote.getBookTitle())
            .quoteId(curationQuote.getQuoteId())
            .author(curationQuote.getAuthor())
            .publisher(curationQuote.getPublisher())
            .imageName(curationQuote.getQuoteImageName())
            .build();
    }

    public static List<CurationItemResponse> toCurationItemDtoListFromQuotes(List<CurationQuote> curationQuotes) {
        Map<Long, List<CurationQuote>> curationQuotesByItemId = curationQuotes.stream()
            .collect(Collectors.groupingBy(CurationQuote::getCurationItemId));

        return curationQuotesByItemId.entrySet().stream()
            .map(entry -> {
                Long curationItemId = entry.getKey();
                List<CurationQuote> curationItems = entry.getValue();

                CurationQuote representativeQuote = curationItems.get(0);

                List<CurationContentResponse> contents = curationItems.stream()
                    .map(CurationMapper::toCurationContent)
                    .toList();

                return new CurationItemResponse(
                    curationItemId,
                    representativeQuote.getTitle(),
                    representativeQuote.getDescription(),
                    CurationType.QUOTE,
                    contents
                );
            })
            .toList();
    }
}

File: GLIM-AOS/app/src/main/java/com/ssafy/glim/core/data/mapper/CurationMapper.kt
package com.ssafy.glim.core.data.mapper

import android.util.Log
import com.ssafy.glim.core.data.dto.response.CurationContentResponse
import com.ssafy.glim.core.data.dto.response.CurationItemResponse
import com.ssafy.glim.core.domain.model.Book
import com.ssafy.glim.core.domain.model.Curation
import com.ssafy.glim.core.domain.model.CurationContent
import com.ssafy.glim.core.domain.model.CurationType
import com.ssafy.glim.core.domain.model.Quote

fun CurationItemResponse.toDomain(): Curation {
    val typeEnum = CurationType.valueOf(this.curationType)
    val books = if (typeEnum == CurationType.BOOK) {
        Log.d("CurationMapper", "Curation type is BOOK, converting contents to Book ${contents.get(0)}")
        this.contents.map { it.toBook() }
    } else {
        emptyList()
    }

    val glims = if (typeEnum == CurationType.QUOTE) {
        this.contents.map { it.toQuote() }
    } else {
        emptyList()
    }

    return Curation(
        id = this.curationItemId,
        title = this.title,
        description = this.description,
        type = typeEnum,
        contents = CurationContent(
            book = books,
            quote = glims
        )
    )
}

private fun CurationContentResponse.toBook(): Book =
    Book(
        bookId = this.bookId ?: -1L,
        title = this.bookTitle,
        author = this.author,
        publisher = this.publisher,
        pubDate = "",
        isbn = "",
        description = "",
        cover = this.bookCoverUrl ?: "",
    )

private fun CurationContentResponse.toQuote(): Quote =
    Quote(
        bookId = bookId ?: -1L,
        isLike = false,
        likes = 0L,
        bookTitle = bookTitle,
        author = author,
        bookCoverUrl = bookCoverUrl ?: "",
        page = 0,
        publisher = publisher,
        quoteId = quoteId ?: -1L,
        quoteImageName = imageName ?: "",
        quoteViews = 0L,
    )
